File: jem-core/src/main/java/jem/util/flob/Flobs.java
/*
 * Copyright 2014-2017 Peng Wan <phylame@163.com>
 *
 * This file is part of Jem.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jem.util.flob;

import jclp.io.IOUtils;
import jclp.io.RAFInputStream;
import jclp.util.Exceptions;
import jclp.vdm.VdmEntry;
import jclp.vdm.VdmReader;
import jem.util.Variants;
import lombok.Getter;
import lombok.NonNull;

import java.io.*;
import java.net.URL;
import java.util.Arrays;
import java.util.zip.ZipFile;

/**
 * Factory class for creating {@code Flob} instance.
 */
public final class Flobs {
    private Flobs() {
    }

    public static Flob forEmpty(String name, String mime) {
        return forBytes(name, null, mime);
    }

    public static Flob forFile(@NonNull File file, String mime) throws IOException {
        return new NormalFlob(file, mime);
    }

    public static Flob forZip(@NonNull ZipFile zipFile, @NonNull String entry, String mime) throws IOException {
        return new EntryFlob(zipFile, entry, mime);
    }

    public static Flob forVam(@NonNull VdmReader vam, @NonNull String entry, String mime) throws IOException {
        return new VamFlob(vam, entry, mime);
    }

    public static BlockFlob forBlock(@NonNull String name, @NonNull RandomAccessFile file, long offset, long size,
                                     String mime) throws IOException {
        return new BlockFlob(name, file, offset, size, mime);
    }

    public static Flob forURL(@NonNull URL url, String mime) {
        return new URLFlob(url, mime);
    }

    public static Flob forBytes(@NonNull String name, byte[] bytes, String mime) {
        return new ByteFlob(name, bytes, mime);
    }

    public static class NormalFlob extends AbstractFlob {
        static {
            Variants.mapClass(NormalFlob.class, Variants.FLOB);
        }

        @Getter
        private final File file;

        private NormalFlob(@NonNull File file, String mime) throws IOException {
            super(mime);
            if (!file.exists()) {
                throw Exceptions.forFileNotFound("No such file: %s", file);
            }
            if (file.isDirectory()) {
                throw Exceptions.forIllegalArgument("Expect file but specified directory: %s", file);
            }
            this.file = file;
        }

        @Override
        public String getName() {
            return file.getPath();
        }

        @Override
        public FileInputStream openStream() throws IOException {
            return new FileInputStream(file);
        }
    }

    private static class VamFlob extends AbstractFlob {

        static {
            Variants.mapClass(VamFlob.class, Variants.FLOB);
        }

        private final VdmReader vdmReader;
        private final VdmEntry vdmEntry;
        private final String name;

        VamFlob(VdmReader vdmReader, String name, String mime) throws IOException {
            super(mime);
            vdmEntry = vdmReader.entryFor(name);
            if (vdmEntry == null) {
                throw Exceptions.forIO("No such entry in vdm: %s", name);
            }
            this.vdmReader = vdmReader;
            this.name = name;
        }

        @Override
        public String getName() {
            return name;
        }

        @Override
        public InputStream openStream() throws IOException {
            return vdmReader.streamFor(vdmEntry);
        }

        @Override
        public String toString() {
            return vdmEntry.toString();
        }
    }

    private static class EntryFlob extends AbstractFlob {
        static {
            Variants.mapClass(EntryFlob.class, Variants.FLOB);
        }

        private final ZipFile zipFile;
        private final String name;

        private EntryFlob(@NonNull ZipFile zipFile, @NonNull String name, String mime) throws IOException {
            super(mime);
            if (zipFile.getEntry(name) == null) {
                throw Exceptions.forIO("No such entry in ZIP: %s", name);
            }
            this.zipFile = zipFile;
            this.name = name;
        }

        @Override
        public String getName() {
            return name;
        }

        @Override
        public InputStream openStream() throws IOException {
            return zipFile.getInputStream(zipFile.getEntry(name));
        }

        @Override
        public String toString() {
            return "zip://" + zipFile.getName() + '!' + super.toString();
        }
    }

    /**
     * File object that represents a block of source file.
     */
    public static class BlockFlob extends AbstractFlob {
        static {
            Variants.mapClass(BlockFlob.class, Variants.FLOB);
        }

        private final String name;
        private final RandomAccessFile file;

        public long offset, size;

        private BlockFlob(String name, RandomAccessFile file, long offset, long size, String mime) throws IOException {
            super(mime);
            if (size > (file.length() - offset)) {
                throw Exceptions.forIllegalArgument("offset(%d) + size(%d) > total(%d)", offset, size, file.length());
            }
            this.name = name;
            this.file = file;
            this.offset = offset;
            this.size = size;
        }

        @Override
        public String getName() {
            return name;
        }

        @Override
        public InputStream openStream() throws IOException {
            file.seek(offset);
            return new RAFInputStream(file, size);
        }

        @Override
        public byte[] readAll() throws IOException {
            file.seek(offset);
            byte[] buf = new byte[(int) size];
            int n = file.read(buf);
            if (n < size) {
                return Arrays.copyOf(buf, n);
            } else {
                return buf;
            }
        }

        @Override
        public long writeTo(OutputStream out) throws IOException {
            file.seek(offset);
            return IOUtils.copy(file, out, (int) size);
        }

        @Override
        public String toString() {
            return String.format("block://%s;offset=%d;size=%d", super.toString(), offset, size);
        }
    }

    public static class URLFlob extends AbstractFlob {
        static {
            Variants.mapClass(URLFlob.class, Variants.FLOB);
        }

        @Getter
        private final URL url;

        private URLFlob(URL url, String mime) {
            super(mime);
            this.url = url;
        }

        @Override
        public String getName() {
            return url.getPath();
        }

        @Override
        public InputStream openStream() throws IOException {
            return url.openStream();
        }

        @Override
        public String toString() {
            return url.toString() + ";mime=" + getMime();
        }
    }

    private static class ByteFlob extends AbstractFlob {
        static {
            Variants.mapClass(ByteFlob.class, Variants.FLOB);
        }

        private final String name;
        private final byte[] buf;

        private ByteFlob(String name, byte[] buf, String mime) {
            super(mime);
            this.name = name;
            this.buf = (buf != null) ? buf : new byte[0];
        }

        @Override
        public String getName() {
            return name;
        }

        @Override
        public InputStream openStream() throws IOException {
            return new ByteArrayInputStream(buf);
        }

        @Override
        public byte[] readAll() throws IOException {
            return Arrays.copyOf(buf, buf.length);
        }

        @Override
        public long writeTo(OutputStream out) throws IOException {
            out.write(buf);
            return buf.length;
        }

        @Override
        public String toString() {
            return "bytes://" + super.toString();
        }
    }
}

File: jem-core/src/main/kotlin/jem/util/Flobs.kt
package jem.util

import jclp.io.copyLimited
import jclp.io.detectMime
import jclp.io.getResource
import jclp.io.clippedStream
import java.io.*
import java.net.URL

interface Flob {
    val name: String

    val mime: String

    fun openStream(): InputStream

    fun writeTo(output: OutputStream) = openStream().copyLimited(output)

    companion object {
        fun of(url: URL, mime: String = ""): Flob = URLFlob(url, mime)

        fun of(file: File, mime: String = ""): Flob = FileFlob(file, mime)

        fun of(data: ByteArray, name: String, mime: String = ""): Flob = ByteFlob(name, data, mime)

        fun of(file: RandomAccessFile, name: String, offset: Long, length: Long, mime: String = ""): Flob {
            return BlockFlob(name, file, offset, length, mime)
        }

        fun of(uri: String, loader: ClassLoader? = null, mime: String = ""): Flob? {
            val url = getResource(uri, loader) ?: return null
            return URLFlob(url, mime)
        }

        fun empty(mime: String = "") = of(ByteArray(0), "_empty_", mime)
    }
}

open class FlobWrapper(private val flob: Flob) : Flob {
    override val name = flob.name

    override val mime = flob.mime

    override fun openStream() = flob.openStream()

    override fun writeTo(output: OutputStream) = flob.writeTo(output)

    override fun equals(other: Any?) = flob == other

    override fun hashCode() = super.hashCode()

    override fun toString() = flob.toString()
}

abstract class AbstractFlob(mime: String) : Flob {
    override val mime by lazy { detectMime(name, mime) }

    override fun toString() = "$name;mime=$mime"
}

private class URLFlob(private val url: URL, mime: String) : AbstractFlob(mime) {
    override val name: String = url.path

    override fun openStream(): InputStream = url.openStream()

    override fun toString() = "$url;mime=$mime"
}

private class FileFlob(private val file: File, mime: String) : AbstractFlob(mime) {
    override val name: String = file.path

    override fun openStream() = file.inputStream()
}

private class ByteFlob(override val name: String, private val data: ByteArray, mime: String) : AbstractFlob(mime) {
    override fun openStream() = ByteArrayInputStream(data)

    override fun writeTo(output: OutputStream): Long {
        output.write(data)
        return data.size.toLong()
    }

    override fun toString() = "bytes://${super.toString()}"
}

internal class BlockFlob(
        override val name: String,
        private val file: RandomAccessFile,
        private val offset: Long,
        private val length: Long,
        mime: String
) : AbstractFlob(mime) {
    init {
        require(offset + length <= file.length()) { "offset($offset) + length($length) > total(${file.length()})" }
    }

    override fun openStream(): InputStream {
        file.seek(offset)
        return file.clippedStream(offset, length)
    }

    override fun writeTo(output: OutputStream): Long {
        file.seek(offset)
        return file.copyLimited(output, length)
    }
}
